core/src/saros/session/internal/SarosSession.java

```diff
@@ -861,4 +877,42 @@ boolean hasActivityProducers() {
   boolean hasActivityConsumers() {
     return !activeActivityConsumers.isEmpty() || !passiveActivityConsumers.isEmpty();
   }
+
+  /**
+   * Returns whether adding the given reference point to the session would create nested reference
+   * points.
+   *
+   * @param addedReferencePoint the reference point to add to the session
+   * @return whether adding the given reference point to the session would create nested reference
+   *     points
+   */
+  private boolean wouldCreateNestedReferencePoint(IReferencePoint addedReferencePoint) {
+    for (IReferencePoint sharedReferencePoint : getReferencePoints()) {
+      if (ReferencePointComparator.areNested(addedReferencePoint, sharedReferencePoint)) {
+        return true;
+      }
+    }
+
+    return false;
+  }
+
+  /**
+   * Logs an error and leaves the session.
+   *
+   * <p>This method should be called to end the running session if a user tried to add a new
+   * reference point that would lead to nested reference points being shared.
+   *
+   * @param addedReferencePoint the newly added reference point
+   * @see #wouldCreateNestedReferencePoint(IReferencePoint)
+   */
+  // TODO show user error message
+  private void leaveSessionDueToNestedReferencePoint(IReferencePoint addedReferencePoint) {
+    log.error(
+        "adding reference point "
+            + addedReferencePoint
+            + " would lead to nested reference points; ending session - shared reference points: "
+            + getReferencePoints());
+
+    sessionContainer.getComponent(ISarosSessionManager.class).stopSession(LOCAL_USER_LEFT);
```

Comment:
Beside the fact that aborting the session is very bad as it should not happen (in former version the project was simply not shared, ok not even better but ...) I highly recommend to not call stopSession inside the Session class.

Reply:
If you want to cancel the session do it async. Fork a thread.

Reply:
IIRC stopSession will wait on existing resource negotiation etc. However this will be called inside a resource negotiation. Guess you will run into a nasty timeout with will continue afterwards in an unstable state.

Reply:
Moved to new thread. Works when starting a session or adding resources to the session. I tested this using Saros/E for the sending and receiving side.

> IIRC stopSession will wait on existing resource negotiation etc.

[`SarosSessionManager.stop()` closes all running negotiations.](https://github.com/saros-project/saros/blob/master/core/src/saros/session/SarosSessionManager.java#L327).